Core/include/Acts/Vertexing/SeedVertexFinder.ipp

```diff
@@ -0,0 +1,656 @@
+                });
+
+      unsigned int threshold = (unsigned int)(triplets_with_planes.size() *
+                                              (1. - m_cfg.removeFraction));
+
+      for (unsigned int tr = threshold + 1; tr < triplets_with_planes.size();
+           ++tr) {
+        const auto& abg = triplets_with_planes.at(tr).first.first;
+        const auto& delta = triplets_with_planes.at(tr).first.second;
+
+        // remove this triplet from A and B
+        A -= 2. * (abg * abg.transpose());
+        B += 2. * delta * abg;
+      }
+
+      // remove all excesive triplets
+      triplets_with_planes.resize(threshold);
+    }
+  }
+
+  return vtx;
+}
+
+template <typename spacepoint_t>
+Acts::Ray3D Acts::SeedVertexFinder<spacepoint_t>::makeRayFromTriplet(
+    const Acts::SeedVertexFinder<spacepoint_t>::Triplet triplet) const {
+  Acts::SymMatrix3 mat;
+  mat.row(0) = Acts::Vector3(triplet.a.x(), triplet.a.y(), triplet.a.z());
+  mat.row(1) = Acts::Vector3(triplet.b.x(), triplet.b.y(), triplet.b.z());
+  mat.row(2) = Acts::Vector3(triplet.c.x(), triplet.c.y(), triplet.c.z());
+
+  Acts::Vector3 mean = mat.colwise().mean();
+  Acts::SymMatrix3 cov = (mat.rowwise() - mean.transpose()).transpose() *
+                         (mat.rowwise() - mean.transpose()) / 3.;
+
+  // "cov" is self-adjoint matrix
+  Eigen::SelfAdjointEigenSolver<Acts::SymMatrix3> saes(cov);
+  // eigenvalues are sorted in increasing order
+  Acts::Vector3 eivec = saes.eigenvectors().col(2);
+
+  return {mean, eivec};
+}
+
+template <typename spacepoint_t>
+Acts::Vector3 Acts::SeedVertexFinder<spacepoint_t>::findClosestPointFromRays(
+    const std::vector<Acts::SeedVertexFinder<spacepoint_t>::Triplet>& triplets)
+    const {
+  // define function f = sum over all triplets [distance from an unknown point
+  // (x_0,y_0,z_0) to the ray defined by the triplet] find minimum of "f" by
+  // partial derivations over x_0, y_0, and z_0 each derivation has parts
+  // lineary depending on x_0, y_0, and z_0 (will fill A[][3]) or to nothing
+  // (will fill B[]) solve A*(x_0,y_0,z_0) = B
+
+  Acts::Vector3 vtx = Acts::Vector3::Zero();
+  Acts::Vector3 vtx_prev{m_cfg.rMaxFar, m_cfg.rMaxFar, m_cfg.maxAbsZ};
+
+  // (start_point, direction), distance
+  std::vector<
+      std::pair<std::pair<Acts::Vector3, Acts::Vector3>, Acts::ActsScalar>>
+      triplets_with_rays;
+  triplets_with_rays.reserve(triplets.size());
+
+  for (const auto& triplet : triplets) {
+    triplets_with_rays.emplace_back(
+        std::make_pair(triplet.ray.origin(), triplet.ray.dir()), -1.);
+  }
+
+  // elements of the linear equations to solve
+  Acts::SymMatrix3 A = Acts::SymMatrix3::Identity() * 2. * triplets.size();
+  Acts::Vector3 B = Acts::Vector3::Zero();
+  for (const auto& triplet : triplets_with_rays) {
+    // use ray saved from earlier
+    const auto& start_point = triplet.first.first;
+    const auto& direction = triplet.first.second;
+
+    A -= 2. * (direction * direction.transpose());
```

Comment:
`squaredNorm`?

Reply:
Hi @andiwand, same as above, we'd like to make a matrix.

Cheers, Petr